File: MSVS/RM.Win.MonitorPower/RM.Win.MonitorPower/PhysicalMonitor.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using RM.Win.MonitorPower.Interop;

namespace RM.Win.MonitorPower
{
	internal sealed class PhysicalMonitor : IDisposable
	{
		public PhysicalMonitor(SafeHandle handle, string description)
		{
			Handle = handle;
			Description = description;
		}

		public SafeHandle Handle { get; }

		public string Description { get; }

		public void SetPowerState(MonitorPowerState powerState)
		{
			Interop.Win32.SetMonitorPower(Handle, (uint)powerState);
		}

		public static IReadOnlyList<PhysicalMonitor> GetPhysicalMonitors()
		{
			return Win32.GetPhysicalMonitors();
		}

		public void Dispose()
		{
			Handle?.Dispose();
		}

		public override bool Equals(object obj)
		{
			return obj is PhysicalMonitor mon && Handle.Equals(mon.Handle);
		}

		public override int GetHashCode()
		{
			return Handle.GetHashCode();
		}

		public override string ToString()
		{
			return $"Physical monitor {Handle}: {Description}";
		}
	}
}

File: MSVS/RM.Win.MonitorPower/RM.Win.MonitorPower.Lib/PhysicalMonitor.cpp
#include "pch.h"
#include "PhysicalMonitor.h"
#include "dxva2.h"

using namespace System::Collections::Generic;

namespace RM
{
	namespace Win
	{
		namespace MonitorPower
		{
			namespace Lib
			{
				PhysicalMonitor::PhysicalMonitor(const IntPtr handle, String^ description)
				{
					_handle = handle;
					_description = description;
				}

				void PhysicalMonitor::Destroy()
				{
					dxva2::destroy_monitor(_handle.ToPointer());
				}

				PhysicalMonitor::!PhysicalMonitor()
				{
					Destroy();
				}

				PhysicalMonitor::~PhysicalMonitor()
				{
					Destroy();
				}

				void PhysicalMonitor::SetPowerState(MonitorPowerState state)
				{
					constexpr BYTE MONITOR_POWER = 0xD6;
					if (!dxva2::set_monitor_vcp_value(_handle.ToPointer(), MONITOR_POWER, safe_cast<DWORD>(state)))
					{
						throw gcnew ComponentModel::Win32Exception(Runtime::InteropServices::Marshal::GetLastWin32Error());
					}
				}

				array<PhysicalMonitor^>^ PhysicalMonitor::GetPhysicalMonitors()
				{
					std::vector<PHYSICAL_MONITOR> monitors;

					if (!dxva2::get_physical_monitors(monitors))
					{
						throw gcnew ComponentModel::Win32Exception(Runtime::InteropServices::Marshal::GetLastWin32Error());
					}
					
					auto result = gcnew List<PhysicalMonitor^>();

					for (auto const& monitor: monitors)
					{
						result->Add(gcnew PhysicalMonitor(IntPtr(monitor.hPhysicalMonitor), gcnew String(monitor.szPhysicalMonitorDescription)));
					}

					return result->ToArray();
				}
			}
		}
	}
}
